tutorials/tour/case-classes.md

```diff
@@ -10,80 +10,125 @@ tutorial-next: pattern-matching
 tutorial-previous: currying
 ---
 
-Scala supports the notion of _case classes_. Case classes are regular classes which export their constructor parameters and which provide a recursive decomposition mechanism via [pattern matching](pattern-matching.html).
+Scala supports the notion of _case classes_. Case classes are just regular classes that are:
 
-Here is an example for a class hierarchy which consists of an abstract super class `Term` and three concrete case classes `Var`, `Fun`, and `App`.
+* Immutable by default
+* Decomposable through [Pattern Matching](pattern-matching.html)
+* Compared by structural equality instead of by reference
+* Succinct to instantiate and operate on
+
+This is achieved through auto-generated code by the Scala compiler.
+
+Here is an example for a Notification type hierarchy which consists of an abstract super class `Notification` and three concrete Notification types implemented with case classes `Email`, `SMS`, and `VoiceRecording`.
 
 ```tut
-abstract class Term
-case class Var(name: String) extends Term
-case class Fun(arg: String, body: Term) extends Term
-case class App(f: Term, v: Term) extends Term
+abstract class Notification
+case class Email(sourceEmail: String, title: String, body: String) extends Notification
+case class SMS(sourceNumber: String, message: String) extends Notification
+case class VoiceRecording(contactName: String, link: String) extends Notification
 ```
 
-This class hierarchy can be used to represent terms of the [untyped lambda calculus](https://en.wikipedia.org/wiki/Lambda_calculus). To facilitate the construction of case class instances, Scala does not require that the `new` primitive is used. One can simply use the class name as a function.
+Instantiating a Case class is easy: (Note that we don't need to use the `new` keyword)
+
+```tut
+val emailFromJohn = Email("john.doe@mail.com", "Greetings From John!", "Hello World!")
+```
 
-Here is an example:
+The constructor parameters of case classes are treated as public values and can be accessed directly. 
 
 ```tut
-Fun("x", Fun("y", App(Var("x"), Var("y"))))
+val title = emailFromJohn.title
+println(title) // prints "Greetings From John!"
 ```
 
-The constructor parameters of case classes are treated as public values and can be accessed directly.
+With case classes, you cannot mutate their fields directly. Instead, you make a copy using the `copy` method.
+As seen below, you can replace just some of the fields:
 
 ```tut
-val x = Var("x")
-println(x.name)
+emailFromJohn.title = "Goodbye From John!" // This is a compilation error. We cannot assign another value to val fields, which all case classes fields are by default.
+
+val editedEmail = emailFromJohn.copy(title = "I am learning Scala!", body = "It's so cool!")
+
+println(emailFromJohn) // prints "Email(john.doe@mail.com,Greetings From John!,Hello World!)"
+println(editedEmail) // prints "Email(john.doe@mail.com,I am learning Scala,It's so cool!)"
```

Comment:
Since this code block is designed to fail then `tut` should be `tut:fail`. This will allow tut to give an error if the code block compiles. It is a little weird.

https://github.com/tpolecat/tut

Also check the Travis logs if they fail. They will show you where to look to fix the problem. :-)

https://travis-ci.org/scala/scala.github.com/builds/166207587